pkg/components/exec_node_base.go

```diff
@@ -3,56 +3,147 @@ package components
 import (
 	"context"
 	"log"
+	"path"
 
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/util/yaml"
 	ptr "k8s.io/utils/pointer"
 
+	ytv1 "github.com/ytsaurus/yt-k8s-operator/api/v1"
+
+	"github.com/ytsaurus/yt-k8s-operator/pkg/consts"
 	"github.com/ytsaurus/yt-k8s-operator/pkg/resources"
 	"github.com/ytsaurus/yt-k8s-operator/pkg/ytconfig"
 )
 
 type baseExecNode struct {
-	server     server
-	cfgen      *ytconfig.NodeGenerator
-	sidecars   []string
-	privileged bool
+	server server
+	cfgen  *ytconfig.NodeGenerator
+	spec   *ytv1.ExecNodesSpec
+
+	sidecarConfig *ConfigHelper
+}
+
+func (n *baseExecNode) IsJobResourcesIsolated() bool {
+	if n.spec.JobEnvironment != nil && n.spec.JobEnvironment.CRI != nil {
+		return true
+	}
+	return false
 }
 
 func (n *baseExecNode) Fetch(ctx context.Context) error {
-	return resources.Fetch(ctx, n.server)
+	return resources.Fetch(ctx, n.server, n.sidecarConfig)
 }
 
-func (n *baseExecNode) doSyncBase(ctx context.Context, dry bool) (ComponentStatus, error) {
-	var err error
-	if !dry {
-		setContainerPrivileged := func(ct *corev1.Container) {
-			if ct.SecurityContext == nil {
-				ct.SecurityContext = &corev1.SecurityContext{}
+func (n *baseExecNode) doBuildBase() error {
+	statefulSet := n.server.buildStatefulSet()
+	podSpec := &statefulSet.Spec.Template.Spec
+
+	// Pour job resources into node container if jobs are not isolated.
+	if n.spec.JobResources != nil && !n.IsJobResourcesIsolated() {
+		addResourceList := func(list, newList corev1.ResourceList) {
+			for name, quantity := range newList {
+				if value, ok := list[name]; !ok {
+					list[name] = quantity.DeepCopy()
+				} else {
+					value.Add(quantity)
+					list[name] = value
+				}
 			}
-			ct.SecurityContext.Privileged = ptr.Bool(n.privileged)
 		}
 
-		statefulSet := n.server.buildStatefulSet()
-		containers := &statefulSet.Spec.Template.Spec.Containers
-		if len(*containers) != 1 {
-			log.Panicf("length of exec node containers is expected to be 1, actual %v", len(*containers))
+		addResourceList(podSpec.Containers[0].Resources.Requests, n.spec.JobResources.Requests)
+		addResourceList(podSpec.Containers[0].Resources.Limits, n.spec.JobResources.Limits)
+	}
+
+	setContainerPrivileged := func(ct *corev1.Container) {
+		if ct.SecurityContext == nil {
+			ct.SecurityContext = &corev1.SecurityContext{}
+		}
+		ct.SecurityContext.Privileged = ptr.Bool(n.spec.Privileged)
+	}
+
+	if len(podSpec.Containers) != 1 {
+		log.Panicf("length of exec node containers is expected to be 1, actual %v", len(podSpec.Containers))
```

Comment:
Number of

Reply:
Done. Not my code, I've just scrambled it. =)